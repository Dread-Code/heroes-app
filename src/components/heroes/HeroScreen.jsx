import React from "react";
import { Redirect, useParams } from "react-router-dom";
import { getHeroesById } from "../../selectors/getHeroesById";

export const HeroScreen = ({ history }) => {
  const { heroId } = useParams();
  const hero = getHeroesById(heroId)[0];

  if (!hero === 0) {
    return <Redirect to="/marvel" />;
  }

  const { superhero, publisher, alter_ego, first_appearance, characters } =
    hero;

  const handleReturn = () => {
    if (history.length <= 2) {
      history.push("/");
    }
    history.goBack();
  };

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`../assets/heroes/${heroId}.jpg`}
          alt={superhero}
          className="img-thumbnail"
        />
      </div>
      <div className="col-8">
        <h3>{superhero}</h3>
        <ul className="list-group list-group.flush">
          <li className="list-group-item">
            <b>Alter ego: </b> {alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publishe: </b> {publisher}
          </li>
          <li className="list-group-item">
            <b>First appearence: </b> {first_appearance}
          </li>
        </ul>
        <h5>Characters</h5>
        <p>{characters}</p>
        <button className="btn btn-outline-info" onClick={handleReturn}>
          return
        </button>
      </div>
    </div>
  );
};
