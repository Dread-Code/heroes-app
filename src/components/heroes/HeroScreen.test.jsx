import React from "react";
import { mount } from "enzyme";
import { HeroScreen } from "./HeroScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("HeroScreen", () => {
  const historyMock = {
    length: 10,
    push: jest.fn(),
    goBack: jest.fn(),
  };

  test("should match snapshoot if there isn't URL arguments", () => {
    const wrapper = mount(
      /**
       * initialEntries
       * Es una propiedad que no ayuda a pasarle argumentos a nuestra ruta
       *  */
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("should render a hero if the parameter it`s find ", () => {
    const wrapper = mount(
      /**
       * Es una propiedad que no ayuda a pasarle argumentos a nuestra ruta
       * AUn asi neceitamos con el useparams obtener el id al cual hace referencia
       * en la linea 6 del componente, y para esto nosotros podemos simular por
       * medio de HOC Route esta query que va a obtener
       *  */
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find(".row").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test("should return the last screen with push", () => {
    const historyMock = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(historyMock.push).toHaveBeenCalledTimes(1);
    expect(historyMock.goBack).toHaveBeenCalledTimes(0);
  });

  test("should go back", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");

    expect(historyMock.push).toHaveBeenCalledTimes(0);
    expect(historyMock.goBack).toHaveBeenCalledTimes(1);
  });

  test("should call redirect if the hero doesn't exist", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/marvel-spider123412341234"]}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );

    expect(wrapper.text()).toBe("");
  });
});
