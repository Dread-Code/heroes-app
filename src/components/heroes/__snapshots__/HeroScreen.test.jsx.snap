// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroScreen should match snapshoot if there isn't URL arguments 1`] = `""`;

exports[`HeroScreen should render a hero if the parameter it\`s find  1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="animate__animated animate__fadeIn col-8"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group.flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego: 
        </b>
         
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publishe: 
        </b>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First appearence: 
        </b>
         
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      return
    </button>
  </div>
</div>
`;
